tuta-sdk/rust/src/crypto/ecc.rs

```diff
@@ -1,14 +1,101 @@
+use rand_core::CryptoRngCore;
 use zeroize::*;
 
-#[derive(Zeroize, ZeroizeOnDrop)]
-pub struct EccPrivateKey([u8; 32]);
+const ECC_KEY_SIZE: usize = 32;
 
-#[derive(Zeroize, ZeroizeOnDrop)]
+#[derive(ZeroizeOnDrop, Clone)]
+pub struct EccPrivateKey([u8; ECC_KEY_SIZE]);
+
+impl EccPrivateKey {
+    /// Get a reference to the underlying bytes.
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_slice()
+    }
+
+    /// Calculate the public key for this private key.
+    pub fn derive_public_key(&self) -> EccPublicKey {
+        use curve25519_dalek::Scalar;
+        use curve25519_dalek::montgomery::MontgomeryPoint;
+
+        // public key = private key * base point
+        let public_key = Zeroizing::new(
+            MontgomeryPoint::mul_base(&*Zeroizing::new(Scalar::from_bytes_mod_order(self.0)))
+        );
+
+        EccPublicKey(public_key.0)
+    }
+
+    /// Attempt to convert a slice of bytes into an ECC key.
+    ///
+    /// Returns `Err` on failure.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, EccKeyError> {
+        match bytes.len() {
+            ECC_KEY_SIZE => Ok(Self(bytes.try_into().unwrap())),
+            actual_size => Err(EccKeyError { actual_size })
+        }
+    }
```

Comment:
```suggestion
    pub fn from_bytes(bytes: &[u8; ECC_KEY_SIZE]) -> Result<Self, EccKeyError> {
        match bytes.len() {
            ECC_KEY_SIZE => Ok(Self(*bytes)),
            actual_size => Err(EccKeyError { actual_size })
        }
    }
```

If the byte slice has to be a set size, why not change the type to include the size instead of panicking?

Reply:
Fair enough, it would be good to note that it is optimized away in a comment.

Reply:
I think the whole idea of `from_bytes` (maybe it should be named `from_slice`) is that we do it only once

Reply:
and we have `arr_cast_size` exactly  for that if you don't want to unwrap it

Reply:
ah no that one casts between known arrays my bad

Reply:
you can also just match on `try_into()` result if you want to be explicit that the check is only done once?